docs: managing runs (#741)
jina-ai/finetuner: docs/concepts/manage-runs.md

(manage-runs)=
# {octicon}`terminal` Managing Runs

> After you have started a fine-tuning or a synthesis job by calling {meth}`~finetuner.fit` or {meth}`~finetuner.synthesize`,
> you get a {class}`~finetuner.run.Run` object in response.
> This run object serves as a handle for the cloud job that you triggered.

Reviewer: ```suggestion
After starting a fine-tuning or a synthesis job by calling {meth}`~finetuner.fit` or {meth}`~finetuner.synthesize`,
those methods return a {class}`~finetuner.run.Run` object. This object holds information about your job and serves as a handle to access the job running on the Jina AI Cloud.

```

> If you lost the run object for the job, you can create a new `Run` object with the {meth}`finetuner.get_run` function:

Reviewer: ```suggestion
For example, the name of your run is in the `name` attribute of the `Run` object:

```python
run = finetuner.fit(
    ...
)

print(run.name)
```

If you end your session or lose track of the `Run` object for the job, you can create a new `Run` object with the {meth}`finetuner.get_run` function:
```

> 

Reviewer: ```suggestion

If you end your session or lose track of the `Run` object for the job, you can create a new `Run` object with the {meth}`finetuner.get_run` function:

```

```python
import finetuner

finetuner.login()

> run = finetuner.get_run('your-run-name', 'your-experiment-name')

Reviewer: ```suggestion
run = finetuner.get_run('your-run-name')
```

```

> Since your runs are only accessible to you, you need to login first.
> If you did not specify an experiment name when calling the `fit` or `synthesize` function, 
> you can omit the experiment name in the `get_run` function as well.

Reviewer: ```suggestion
Your runs are only accessible to you, so you need to log in first.

If you used the `experiment_name` parameter in your call to {meth}`~finetuner.fit` or {meth}`~finetuner.synthesize` to specify an experiment name, 
you will need to add that to the `get_run` call:

```python
run = finetuner.get_run('your-run-name', 'your-experiment-name')
```
```

> Further, you can use the {meth}`finetuner.list_runs` function to obtain a list of run objects for all jobs you submitted.

Reviewer: ```suggestion
Furthermore, you can use the {meth}`finetuner.list_runs` function to list the names and experiment names of all the jobs you have submitted. If you have lost all information about your run, you can look for it in that list.
```

Author: it does not list the names but it returns a list of run objects.

```python

> run = finetuner.get_run('your-run-name', 'your-experiment-name')

Reviewer: ```suggestion
run = finetuner.get_run('your-run-name', 'your-experiment-name')
```
```

If this list is very long, the function call might take a while since it needs to send multiple requests.
```

## Attributes of a run

A `Run` object has the following attributes:
- `name`: It references a job in the Jina AI Cloud. Therefore, it has to be unique for its experiment. If you don't specify a name yourself, Finetuner will give it a memorable name.
- `config`: A configuration which is serialized from the attributes, which you passed to the `fit` or `synthesize` function.

> - `artifact_id`: The outcome of your job is stored in an artifact on Jina AI Cloud. This attribute holds its id. You can use this artifact id for instance in the {meth}`~finetuner.get_model` function to retrieve the model for inference as explained in the [Inference Section](inference).

Reviewer: ```suggestion
- `artifact_id`: The results of your job are stored in an artifact on Jina AI Cloud identified by this artifact id. You can use this id in the {meth}`~finetuner.get_model` function to retrieve your fine-tuned mode, as explained in the [Inference Section](inference).
```

## Check the status of a run
You can check the status of a `Run` via the {meth}`~finetuner.run.Run.status` method:

```python
print(f'Run status: {run.status()}')
```

> You'll see something like this in the terminal:

Reviewer: ```suggestion
If you just launched the job, you'll likely see something like this in the terminal:
```

```bash
Run status: CREATED
```

During fine-tuning or synthesis, the run status changes from:

> 1. CREATED: The {class}`~finetuner.run.Run` has been created and submitted to the job queue.

Reviewer: I think I proposed that too.

2. STARTED: The job is in progress.
3. FINISHED: The job finished successfully and the model has been sent to Jina AI Cloud.
4. FAILED: The job failed, please check the logs for more details.

## Retrieve the logs of a run

> When the status of your run is STARTED or FINISHED,

Reviewer: ```suggestion
If the status of your run is STARTED or FINISHED,
```

> you can retrieve logs from the cloud job with the {meth}`~finetuner.run.Run.logs` method.
> This will return a list of all log messages produced by the job so far.

Reviewer: ```suggestion
you can retrieve the job's logs from Jina AI Cloud with the {meth}`~finetuner.run.Run.logs` method.
This will return a list of all log messages produced by the job so far.
```

> Alternatively, you can streamline the logs to your console.
> Therefore, Finetuner provides the {meth}`~finetuner.run.Run.stream_logs` method.
> Executing the following code will wait until your job obtains the status STARTED 
> if it is still CREATED and continuously prints logs to your console when they come in until the job finished:

Reviewer: ```suggestion
You can stream the logs to your console with the {meth}`~finetuner.run.Run.stream_logs` method.
If your run's status is `CREATED`, this method will wait until your job status is `STARTED` and then continuously prints logs to your console and continues until your job has either finished or failed
```

```python
for entry in run.stream_logs():
    print(entry)

> ```

Author: no we don't have a CLI only the Python client and the Web Interface. There you can also see the logs.

> ```

Author: you could query the logs via a rest client, e.g, using curl, wget, etc.

> Besides log messages, you can obtain more tracking information with the WandBLogger callback.
> Check out the [Callbacks Section](callbacks) to implement this.

Reviewer: ```suggestion
In addition to log messages, you can obtain more tracking information with the `WandBLogger` callback.
Check out the [Callbacks Section](callbacks) to implement this.
```

> ## View runs in Jina AI Cloud
> 
> You can also view your runs in the Jina AI Cloud.
> After you login (see [Login Section](login)),
> you can click on the "Finetuner" section in the left sidebar at the [cloud.jina.ai](https://cloud.jina.ai/) Website.
> This will show you a list of all your runs:

Reviewer: ```suggestion
## Viewing runs on the Jina AI Cloud web console

You can also view your runs in your browser via the Jina AI Cloud website.
After you log in via your browser at [cloud.jina.ai](https://cloud.jina.ai/) (see [Login Section](login)),
click the "Finetuner" section in the left sidebar to see a list of all your runs:
```

Author: maybe it is confusing to reference the login page here which mainly explains the Python login method which is not related to this.

Author: ok thank you and sorry, the reference to the login page was my mistake.

![Jina AI Cloud - List of Runs](https://user-images.githubusercontent.com/6599259/233099591-d27405b3-a26c-4951-81df-2c5dc096113e.png)

You will see more details of a specific run after you click on the respective row.
It will show you information, like the logs and the configuration of the job:

![Jina AI Cloud - View Logs](https://user-images.githubusercontent.com/6599259/233099603-6af406e1-15c1-401b-af5a-495404114f4c.png)

> ## Save artifacts

Reviewer: ```suggestion
## Saving artifacts
```

> When a job finishes, it produces a model (for fine-tuning runs) or a training dataset (for synthesis runs).

Reviewer: ```suggestion
When a fine-tuning job finishes, it stores a model or models, and for synthesis jobs, it stores a new training dataset.
```

> In both cases, its output is stored in an artifact on Jina AI Cloud.
> You can save this artifact with the {meth}`~finetuner.run.Run.save_artifact` method:

Reviewer: ```suggestion
In either case, you can save this result with the {meth}`~finetuner.run.Run.save_artifact` method:
```

```python
run.save_artifact('path/for/tuned_model')
```

> To see how to use this model, read further in the [Inference Section](inference).

Reviewer: ```suggestion
The [Inference Section](inference) describes how to use downloaded models. 
```

## Evaluation data

> Also worth mentioning is that a `Run` object has methods like {meth}`~finetuner.run.Run.metrics` and
> {meth}`~finetuner.run.Run.example_results` to retrieve results produced by an `EvaluationCallback`.
> Details are provided in the [Evaluation Section](evaluation).

Reviewer: ```suggestion
The `Run` object also has the methods {meth}`~finetuner.run.Run.metrics` and
{meth}`~finetuner.run.Run.example_results`, which retrieve results produced by an `EvaluationCallback`.
Details are provided in the [Evaluation Section](evaluation).
```

```python
import finetuner

finetuner.login()

run = finetuner.get_run('your-run-name')
```
Your runs are only accessible to you, so you need to log in first.

If you used the `experiment_name` parameter in your call to {meth}`~finetuner.fit` or {meth}`~finetuner.synthesize` to specify an experiment name, 
you will need to add that to the `get_run` call:

```python
run = finetuner.get_run('your-run-name', 'your-experiment-name')
```

Furthermore, you can use the {meth}`finetuner.list_runs` function to get a list of run objects for all jobs you submitted (or for a specific experiment).
This might also be helpful, if you have lost all information about your run.

```{admonition} Run-time of the list_runs function
:class: hint
If this list is very long, the function call might take a while since it needs to send multiple requests.
```

## Attributes of a run

A `Run` object has the following attributes:
- `name`: It references a job in the Jina AI Cloud. Therefore, it has to be unique for its experiment. If you don't specify a name yourself, Finetuner will give it a memorable name.

> - `config`: A configuration which is serialized from the attributes, which you passed to the `fit` or `synthesize` function.

Reviewer 2: ```suggestion
- `config`: A configuration which is created from the arguments you passed to the `fit` or `synthesize` function.
```
